ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java

```diff
@@ -618,9 +672,30 @@ void handleBody(InputStream payload, MessageMetadata messageMetadata) throws IOE
       if (!successfullyDeserialized) {
         BlobData blobData = MessageFormatRecord.deserializeBlob(payload);
         ByteBuffer encryptionKey = messageMetadata == null ? null : messageMetadata.getEncryptionKey();
-        // @todo Use the above encryption key for decryption of blobData.
-        chunkIndexToBuffer.put(chunkIndex, filterChunkToRange(blobData));
-        numChunksRetrieved++;
+        if (encryptionKey == null) {
+          chunkIndexToBuffer.put(chunkIndex, filterChunkToRange(blobData.getStream().getByteBuffer()));
+          numChunksRetrieved++;
+          blobContentAvailableToProcess = true;
+        } else {
+          state = ChunkState.Decrypting;
+          logger.trace("Marking the chunk's state to {} and submitting decrypt job for data chunk {}",
+              ChunkState.Decrypting, chunkBlobId);
+          exec.submitJob(
+              new DecryptJob(chunkBlobId, encryptionKey, blobData.getStream().getByteBuffer(), null, cryptoService, kms,
+                  (DecryptJob.DecryptJobResult result, Exception exception) -> {
+                    logger.trace("Handling decrypt job callback for data chunk {}", chunkBlobId);
+                    if (chunkBlobId.equals(result.getBlobId())) {
+                      decryptCallbackResult = result;
+                      decryptCallbackException = exception;
+                      decryptCallbackResultAvailable = true;
+                      logger.trace("Successfully stored decrypt job results for data chunk {}", chunkBlobId);
+                    } else {
+                      // todo add metric.
```

Comment:
Yeah, initially I thought it could happen. Bcoz, GetChunks will get repurposed for future chunks. But that will happen only if everything succeeds. If there is a failure(which could happen from the response from a diff server for the same chunk), the operation might fail completely. So, we can disregard this check. May be I will set InternalServerError if this happens(even though it shouldn't)